field/TinyLang/Field/Typed/TypeChecker.hs

```diff
@@ -44,107 +33,94 @@ import qualified TinyLang.Field.Typed.Core  as T
     deriving newtype ( Monad
                      , Functor
                      , Applicative
-                     , MonadError e
+                     , MonadError TypeCheckError
                      , MonadSupply
-                     , MonadScope
+                     , MonadReader TyEnv
                      )
 
 {-| A simple type checker
 -}
-type TypeChecker = TypeCheckerT TypeCheckError Identity
+type TypeChecker = TypeCheckerT Identity
 
 {-| Run a type checker function. Note that if there are several variables with the same textual name
 then the resulting scope will only contain the last one.
 -}
-runTypeChecker :: (MonadError TypeCheckError m, MonadSupply m) => TypeChecker a -> m (Scoped a)
+runTypeChecker :: (MonadError TypeCheckError m, MonadSupply m) => TypeChecker a -> m a
 runTypeChecker typeChecker =
-    liftSupply (runStateT (runExceptT $ runTypeCheckerT typeChecker) mempty)
-        >>= \(errOrRes, scope) -> Scoped scope <$> liftEither errOrRes
+    liftSupply ( flip runReaderT mempty
+               . runExceptT
+               . runTypeCheckerT
+               $ typeChecker
+               )
+        >>= liftEither
 
 {-|
 -}
-typeCheck
-    :: (MonadError TypeCheckError m, MonadSupply m, TextField f)
-    => R.Expr R.Var f -> m (Scoped (T.SomeUniExpr f))
-typeCheck = runTypeChecker . inferExpr
-
-{-|
--}
-typeProgram
-    :: (MonadError TypeCheckError m, MonadSupply m, TextField f)
-    => R.Program R.Var f -> m (Scoped (T.Program f))
+typeProgram ::
+    (MonadError TypeCheckError m, MonadSupply m, TextField f)
+    => R.Program R.Var f -> m (T.Program f)
 typeProgram = runTypeChecker . checkProgram
 
-{-|
+{-| Add a variable to type environment
 -}
-checkType
-    :: (MonadError TypeCheckError m, MonadSupply m, TextField f, KnownUni f a)
-    => R.Expr R.Var f -> m (Scoped (T.Expr f a))
-checkType = runTypeChecker . checkExpr
-
-{-| Look up a variable name. If we've already seen it, return the corresponding Var;
-otherwise, increase the Unique counter and use it to construct a new Var.
--}
-makeVar :: (MonadSupply m, MonadScope m) => String -> m Var
-makeVar name = do
-    vars <- get
-    case Map.lookup name vars of
-        Just var -> pure var
-        Nothing  -> do
-            var <- freshVar name
-            put $ Map.insert name var vars
-            pure var
-
-mkSomeUniVar :: forall f. T.Var -> T.SomeUniVar f
-mkSomeUniVar var
-    | '?':_ <- _varName var = Some $ T.UniVar Bool   var
-    | '#':_ <- _varName var = Some $ T.UniVar Vector var
-    | otherwise             = Some $ T.UniVar Field  var
+-- NOTE:  At the moment this mimics the old scope
+-- TODO:  Discuss wether we should add an env on top of @withVar@ or not.
+withVar :: (Monad m) => R.Var -> forall r. (T.Var -> TypeCheckerT m r) -> TypeCheckerT m r
+withVar var kont = do
+    tyEnv <- ask
+    case Map.lookup var tyEnv of
+        Just tVar -> kont tVar
```

Comment:
> which is obscured by the fact that we recover the correct scoping in the renamer.

Hm, I looked into the renamer and I don't think it can recover scoping. I'm really confused now, how is this whole thing working? Golden tests suggest that we do have correct scoping, but looking at the code I don't see how that can possibly be the case. Can we have a unit test with 

```
let x = 0;
let x = 1;
assert x == 1;
```

where we only type check things and don't rename them?

Reply:
Oh, I think I see what's going on. So taking

```
let x = 0;
let x = 1;
assert x == 1;
```

we call

```haskell
checkStatement (R.ELet var m) kont =
    case uniOfVar . R.unVar $ var of
        Some uni -> do
                tM <- T.withKnownUni uni $ checkExpr m
                withVar var $ \ tVar -> kont [T.ELet (UniVar uni tVar) tM]
```

over the first and then the second let, so we extend the initially empty environment first, but then we don't extend it and instead use the already existing binding. So the program elaborates to

```
let x_0 = 0;
let x_0 = 1;
assert x_0 == 1;
```

which then gets renamed correctly.

I'm not entirely sure if it should be causing us any problems right now, but it certainly will in future. For example, we'll need to inline the value of a loop variable and if the loop variable has the same name than some previously referenced variable, then we can't just use the value of that variable, we have to actually update the environment instead of reusing the existing entry that just became shadowed. Or in future we may have stuff like

```
let x = T;
let x = 1;
assert x == 1;
```

and here we also need to update the environment and can't reuse the existing entry, because the type of the new binding differs from the type of the old one.

Reply:
Ok, this actually gave me slight headache as well, because there are two notions of environment conflated into one:
- tracking which uniques are assigned to which identifier, and
- tracking which variables are in scope.

I do not think that untangling them together should be an issue.

# Explanation

1. I refactored `withVar` to pass the typed version of the variable to the continuation in order to encapsulate the process of changing from `R.Var` to `T.Var`.
2. To keep the old behaviour of giving the same unique to the same identifier, we use a `Map R.Var T.Var` to track _which uniques are assigned to which identifiers_.
3. We reuse the same map to track which variables are in scope.
4. As variables are intrinsically typed, we use the name of variable to infer its type/uni.

Once we switch to extrinsically-typed variables, we will probably need to:
- Use a different map for tracking which uniques have been assigned (`Map (R.Var, Uni) T.Var`)
- Use a proper typing environment, which similar to `Map R.Var (T.Var, Uni)`.

Reply:
Previously we had this:

> tracking which uniques are assigned to which identifier, and

and not this:

> tracking which variables are in scope.

Now that we have the latter and given that I can always get the list of external variables of a program, I don't think we need the former anymore. So once we switch to extrinsically-typed variables I propose to only have `Map R.Var (T.Var, Uni)` (or `Map R.Var T.UniVar` for that matter).

Or do you have a use case for `Map (R.Var, Uni) T.Var`?

Reply:
> Or do you have a use case for `Map (R.Var, Uni) T.Var`?
Not really.

Given that we relax our requirement for uniques, maybe we should parameterise the typed ast by the type of variable as well.  Personally I would like to work without uniques and supply.  Especially, since we will perform a renaming step in the compiler anyways.

Reply:
This still does not look right.

Why do we need the 

```haskell
    case Map.lookup var tyEnv of
        Just tVar -> kont tVar
```

part? We needed this kind of logic previously, but I'm currently failing to see why we need it now.

What happens if we remove that and do

```haskell
            tVar <- freshVar . R.unVar $ var
            local (Map.insert var tVar) $ kont tVar
```

(the `Nothing` case`) unconditionally?